New certificate schema, mandatory type, chart sha and timestamp

mmulholla opened the pull request:
This PR covers several things

1. New schema implemented. See "Certificate" struct in chartverifier/Certificate.go 
2. Added ability to mark checks as Mandatory, Optional or Experimental. All current checks are mandatory
3. Generating a sha256 value for the chart. Basically sort all files into alphabetic order and then create sha from their content. Test case repeatedly creates sha calues for test charts making sure they are always the same.
4. Report is output to console and no longer a file. Running locallly use "command 2> report.yaml", running docker image just "command > report.yaml"
5. Default output is now yaml. 
6. Some tests had to be updated for the new changes and schema

pedjak commented on pkg/chartverifier/certificate.go:
```diff
@@ -16,76 +16,88 @@
-import "strconv"
+import (
+	"github.com/redhat-certification/chart-verifier/pkg/chartverifier/checks"
+	helmchart "helm.sh/helm/v3/pkg/chart"
```
is it really needed to introduce package alias? IMHO `chart.Foo` reads perfectly well.

mmulholla replied:
I find "chart" too general, as this is actually the helm chart and I think the alias makes the code more clear.

pedjak replied:
ok

pedjak commented on pkg/chartverifier/certificate.go:
```diff
@@ -16,76 +16,88 @@
-	Name    string `json:"name" yaml:"name"`
-	Version string `json:"version" yaml:"version"`
-}
+var CertificateApiVersion = "v1"
```
is this not actually `ReportApiVersion`? The tool does not generate certificates, it runs checks and report them, IMHO? BTW, I would extract Report type and corresponding types to a separate package, i.e. `pkg/chartverifier/api`?

mmulholla replied:
Use of certificate instead of report goes way beyond this PR. Have opened a new issue for that work: https://github.com/redhat-certification/chart-verifier/issues/78

pedjak commented on pkg/chartverifier/certificate.go:
```diff
@@ -16,76 +16,88 @@
+}
 
-	return report
+func (c *Certificate) IsOk() bool {
```
did not we decide that the tool is not going to aggregate/reason about the checks, i.e. what are mandatory and what not?

mmulholla replied:
Yes, I did this as a helper function for the tests.

pedjak replied:
can you move it then under `certificate_test.go` file, or at least make it private?

mmulholla replied:
I would ask to defer on this to #78 also. The method was not introduced by this PR and is the only method defined in a Certificate interface. To remove the interface creates knock on effects in other places and I can clean it up in #78.

pedjak commented on pkg/chartverifier/certificatebuilder.go:
```diff
@@ -37,62 +39,89 @@ type CheckResult struct {
 
-	if r.ChartVersion == "" {
-		return nil, errors.New("chart version must be set")
+func (by By) Sort(files []*helmchart.File) {
```
can it be private?

mmulholla replied:
Fixed

pedjak commented on pkg/chartverifier/checks/registry.go:
```diff
@@ -42,21 +42,28 @@ func (r *Result) SetResult(outcome bool, reason string) Result {
-	Get(name string) (CheckFunc, bool)
-	Add(name string, checkFunc CheckFunc) Registry
+	Get(name string) (Check, bool)
+	Add(name string, check Check) Registry
```
could `Check` struct have `name` field, so that `Add` can have only single param,  e.g. `Add(check Check)`?

mmulholla replied:
Added. This must be a go v Java thing :-), I still think in Java lol

pedjak commented on pkg/chartverifier/helmcertifier.go:
```diff
@@ -31,9 +31,5 @@ type CertifierBuilder interface {
-
-type Certificate interface {
-	IsOk() bool
+	Certify(uri string) (*Certificate, error)
```
it looks strange to me that `Certificate` can certify something. I would expect that a certificate is result of some done certification. BTW, I would stop calling this certificate and start using Report or something else.

mmulholla replied:
Use of certificate instead of report goes way beyond this PR. Have opened a new issue for that work: #78

mmulholla commented:
@isutton @pedjak Some of the comments relate to use of certificate instead of report. I used certificate because it is consistent with the rest of the code. The schema is in Cerificate.go, it is created from certificatebuilder.go, the  check process is created by certifierbuilder.go and run by certifier.go. If we use Report for the schema we should change all of those, the object and funcs too. Thoughts?

pedjak commented:
> If we use Report for the schema we should change all of those, the object and funcs too. Thoughts?

Yes, change everything what is needed.

mmulholla commented on pkg/chartverifier/certificate.go:
```diff
@@ -16,76 +16,88 @@
-	Ok             bool           `json:"ok" yaml:"ok"`
-	Metadata       *metadata      `json:"metadata" yaml:"metadata"`
-	CheckResultMap checkResultMap `json:"results" yaml:"results"`
+type CheckReport struct {
```
CheckReport is a report from a check. Results is a collection of CheckReport(s), one for each check. Report covers all of the checks and the metadata.

mmulholla commented on pkg/chartverifier/certificate.go:
```diff
@@ -16,76 +16,88 @@
+	newCheck := CheckReport{}
+	newCheck.Check = checkName
+	newCheck.Type = checkType
+	newCheck.Outcome = PassOutcomeType
```
Added.

mmulholla commented on pkg/chartverifier/certificatebuilder.go:
```diff
@@ -37,62 +39,89 @@ type CheckResult struct {
-	return &certificateBuilder{
-		CheckResultMap: checkResultMap{},
-	}
+	cB := certificateBuilder{}
```
Fixed

mmulholla commented on pkg/chartverifier/certificate.go:
```diff
@@ -16,76 +16,88 @@
-	Version string `json:"version" yaml:"version"`
-}
+var CertificateApiVersion = "v1"
+var CertificateKind = "verify-report"
```
Use of certificate instead of report goes way beyond this PR. Have opened a new issue for that work: #78

mmulholla commented on pkg/chartverifier/certificate.go:
```diff
@@ -16,76 +16,88 @@
+	PassOutcomeType OutcomeType = "PASS"
+)
+
+type Certificate struct {
```
Use of certificate instead of report goes way beyond this PR. Have opened a new issue for that work: #78

mmulholla commented on pkg/chartverifier/certificate.go:
```diff
@@ -16,76 +16,88 @@
-type metadata struct {
-	RunMetadata   runMetadata   `json:"tool" yaml:"tool"`
-	ChartMetadata chartMetadata `json:"chart" yaml:"chart"`
+type CertificateMetadata struct {
```
Use of certificate instead of report goes way beyond this PR. Have opened a new issue for that work: #78

mmulholla commented:
@pedjak @isutton  : have addressed the comments which are not connected to using report instead of certificate - for those comments and the entire rename effort I opened a new issue #78
